src/arc.rs

```diff
@@ -183,47 +211,46 @@ impl PathArc {
     /// [`canonicalize`]: struct.PathAbs.html#method.canonicalize
     /// [`current_dir`]: fn.current_dir.html
     pub fn absolute(&self) -> Result<PathAbs> {
-        let mut components = self.components();
-        let mut stack: Vec<OsString> = Vec::new();
-
-        macro_rules! pop_stack { [] => {{
-            if let None = stack.pop() {
-                return Err(Error::new(
-                    io::Error::new(io::ErrorKind::NotFound, ".. consumed root"),
-                    "resolving absolute",
-                    self.clone(),
-                ));
-            }
-        }}}
-
-        handle_prefix(self, &mut stack, &mut components, false)?;
-
-        for component in components {
-            match component {
-                Component::CurDir => { /* ignore, probably impossible */ }
-                Component::Prefix(_) => unreachable!(),
-                Component::RootDir => {
-                    if cfg!(unix) {
-                        unreachable!("root is already handled on unix");
-                    }
-                    // This is actually possible on windows because root is distinct
-                    // from prefix (?)
-                    stack.push(to_os(component));
+        let cwd = env::current_dir().map_err(|e| {
+            Error::new(
+                e,
+                "getting current_dir while resolving absolute",
+                self.clone(),
+            )
+        })?;
+        let mut res = cwd.canonicalize().map_err(|e| {
+            Error::new(e, "canonicalizing", PathArc::new(&cwd))
+        })?;
```

Comment:
One issue with this code is that the doc comment suggests no syscalls will be made if the path is already absolute, while the new implementation always starts by calling `current_dir()` and `canonicalize()`, even if it isn't strictly necessary.

I could change the code to *not* initialize `res` to the canonical current directory at first, and instead conditionally initialize it in the `CurDir`, `ParentDir` and `Normal` arms. The `RootDir` arm would need to conditionally initialize, but only on Windows. That would make this function quite a bit more complex, but it would save on syscalls. Do you think it's worth it?

Reply:
I've pushed a new commit that only gets the current directory if it's actually needed.